### DISCUSSION TITLE
when clause for currently active tab in treeview

### DISCUSSION THREAD

**Discussion by McSodbrenner** [microsoft/vscode-discussions#153] [Extension Development QnA] [answered]
I have a list of all tabs in a treeview and I want to display a certain command ("Pin tab") only for the currently active tab in the context menu of the treeview item. However, I can't find a suitable when clause (https://code.visualstudio.com/api/references/when-clause-contexts). However, I have a feeling that I just don't see it.

**Comment by alefragnani** [comment]
Hi @McSodbrenner ,

I guess you should be looking for the `activeEditor` kind of clauses, like `activeEditor`, `activeEditorGroupIndex` and `activeEditorIsPinned`.

Otherwise, you could create your won `when` clause, using the `setContext` command. More details here: https://code.visualstudio.com/api/references/when-clause-contexts#add-a-custom-when-clause-context

Hope this helps

**Comment by McSodbrenner** [comment, marked as answer]
Hi @alefragnani ,

thank you very much for your answer. My problem is more the value of the other side (the side of the TreeView item) of the comparison.

```
"menus": {
    "view/item/context": [
        {
          "command": "betterOpenEditors.pinTab",
          "when": "view == betterOpenEditors && viewItem == file && activeEditor === ???"
        },
        ...
```

The only feature to address a specific treeview item seems to be the `contextValue` of an item, but which I use to distinguish whether a folder or a file was clicked in my TreeView.
`activeEditor` probably corresponds to a `resourceUri` of a TreeView item (not sure about that), but I don't know how to pass it in the when-clause.

Unfortunately there is no API to pin an inactive tab, hence this workaround to address a specific item only.

**Reply by alefragnani** [reply]
Based on the default keyboard shortcuts (available via `Open Default Keyboard Shortcut (JSON)` command), it seems the `activeEditor` doesn't refer to a `resourceUri`, but instead, the _kinds of editors_ that VS Code supports:

```json
{ "key": "enter",                 "command": "editor.action.extensioneditor.findNext",
                                     "when": "webviewFindWidgetFocused && !editorFocus && activeEditor == 'workbench.editor.extension'" },

{ "key": "enter",                 "command": "editor.action.webvieweditor.findNext",
                                     "when": "webviewFindWidgetFocused && !editorFocus && activeEditor == 'WebviewEditor'" },

{ "key": "shift+alt+f",           "command": "notebook.formatCell",
                                     "when": "editorHasDocumentFormattingProvider && editorTextFocus && inCompositeEditor && notebookEditable && !editorReadonly && activeEditor == 'workbench.editor.notebook'" },
```

So, my guess is, if you need to know which file is focused (equivalent to `activeTextEditor`), you should create your own ContextKey

Hope this helps

**Reply by McSodbrenner** [reply]
OK, the problem seems to be that we only have access to the `contextValue` of a TreeView item via `viewItem`. Since `contextValue` is a string, we are relatively limited in what we can do.
For now I have solved it so that a file item can have three contextValues: `file`, `fileCurrent` and `fileCurrentPinned`.
For all other commands I have now registered an own context, so that I always have the appropriate command:
```
vscode.commands.executeCommand('setContext', 'betterOpenEditors.fileLikeItems', [
	'file',
	'fileCurrent',
	'fileCurrentPinned'
]);
```
package.json
```
{
	"command": "betterOpenEditors.whateverCommand",
	"group": "special",
	"when": "view == betterOpenEditors && viewItem in betterOpenEditors.fileLikeItems"
},
```

For the "Pin tab" and "Unpin tab" command I defined the following:
```
{
	"command": "betterOpenEditors.pinTab",
	"group": "specialPin",
	"when": "view == betterOpenEditors && viewItem == fileCurrent"
},
{
	"command": "betterOpenEditors.unpinTab",
	"group": "specialPin",
	"when": "view == betterOpenEditors && viewItem == fileCurrentPinned"
}
```

Of course I have to care about setting the correct `contextValue` for all TreeView items. And I don't like the solution very much because it's not very flexible if I have to include more similar cases.

And already the next problem comes around the corner.
When I pin an item, everything works so far, but the context menu item does not change to "Unpin tab". I first have to switch to another item and back again, so that the context menu is updated. Apparently the when clauses are not interpreted when the context menu is called, but only when an item is selected. This makes it unusable. Or does anyone have an idea how I could work around this problem?

Thanks for your time so far!

**Reply by alefragnani** [reply]
Well, the amount of contextValues may vary and is up to you how complex your scenarios are. Based on your descriptions, it seems you are in the right direction. And indeed, its flexibility requires you to duplicate some entries in `package.json` for every  scenario.

About your _problem around the corner_, I guess you are missing `onDidPin`/`onDidUnpin` events, where you should call  your `setContext` command again, and update the state of your items. 

Hope this helps

**Reply by r3m0t** [reply]
@McSodbrenner did you fire `onDidChangeTreeData` after updating `TreeItem.contextValue`? As it's needed.

**Reply by McSodbrenner** [reply]
Many thanks to both of you. I went back to work and was able to solve the problem. Every now and then my refresh of the DataProvider was not executed. The solution is not very elegant, but it works for now.

**Comment by r3m0t** [comment]
Using the commands `Developer: Toggle Developer Tools` then `Developer: Inspect Context Keys`, I think you want `resource`, `resourcePath`, `resourceScheme`

**Reply by McSodbrenner** [reply]
Thanks a lot. The problem seems to be that we are only able to access the `contextValue` of a TreeView item (see https://github.com/microsoft/vscode-discussions/discussions/153#discussioncomment-3823074). Unfortunately, the resource* context keys refer to the currently active editor and not to the TreeView element.
